### PR TITLE
Added 'hello-world' quickstart for SDKs

### PR DESCRIPTION
**Pull request by dmitsh**
# Description

Added "Hello World" quickstart for SDK

## Issue reference

We strive to have all PRs being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ x] The quickstart code compiles correctly
* [ x] You've tested new builds of the quickstart if you changed quickstart code
* [ x] You've updated the quickstart's README if necessary

### CODE REVIEW

**Comment by dmitsh** [comment]
@artursouza @yaron2 plz take a look

**Review by wcs1only — COMMENTED** [review]
A couple of general comments:

We have a an automation framework called mechanical-markdown that can run your quickstart as if it were code, and validate output. This helps us avoid having to manually run the quickstarts every release to ensure we haven't broken anything. See: 
https://github.com/dapr/mechanical-markdown/tree/main/examples and the raw README.md for the other quickstarts for examples.

I like using the SDK for the quickstarts. I think perhaps it would be good to replace an existing quickstart with this approach, or perhaps update all of our quickstarts to use the SDKs

File: hello-world-sdk/README.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+expected_stderr_lines:
+output_match_mode: substring
+name: "order get"
+background: true
+sleep: 5
+-->
+
+    ```bash
+    dapr run --app-id order-app --dapr-grpc-port 3500 ./order get
+    ```
```
Comment (halspang) [inline hello-world-sdk/README.md:138]: For commands that have outputs in the quickstarts I think it's good to have an example of what it should look like. We anyone who is going through these to feel confident/be able to verify what they're seeing.

**Review by halspang — COMMENTED** [review]
Looks good to me overall, just think we should add some output examples. 

I also agree with Charlie, this would be great to get into some other quickstarts too. They can be separate PRs but the distributed calculator would be a great one as it can show multiple SDKs.

**Review by artursouza — CHANGES_REQUESTED** [review]
We have decided to keep SDK specific "quickstarts" in the corresponding SDK repo. Please, move this to the go-sdk repo.

@orizohar @AaronCrawfis  thoughts?

**Comment by dmitsh** [comment]
> We have decided to keep SDK specific "quickstarts" in the corresponding SDK repo. Please, move this to the go-sdk repo.
> 
> @orizohar @AaronCrawfis thoughts?

When discussed with @wcs1only and @halspang it seemed there is a need to simplify the quickstarts - to use SDKs instead of Dapr API. I also share this idea. I think we should have a single and simple quickstarts repository.
In this PR I provide list of SDKs where we have more advanced examples.

**Comment by artursouza** [comment]
> > We have decided to keep SDK specific "quickstarts" in the corresponding SDK repo. Please, move this to the go-sdk repo.
> > @orizohar @AaronCrawfis thoughts?
> 
> When discussed with @wcs1only and @halspang it seemed there is a need to simplify the quickstarts - to use SDKs instead of Dapr API. I also share this idea. I think we should have a simple gateway quickstarts repository, to expedite user's adoption.
> In this PR I also provide a full list of language-specific SDKs where we have more advanced examples.

Changing the quickstarts from using APIs to SDKs must be agreed within the maintainers since that is a change in direction we have decided until now.

@AaronCrawfis @orizohar @yaron2  Thoughts?

**Comment by yaron2** [comment]
> > > We have decided to keep SDK specific "quickstarts" in the corresponding SDK repo. Please, move this to the go-sdk repo.
> > > @orizohar @AaronCrawfis thoughts?
> > 
> > 
> > When discussed with @wcs1only and @halspang it seemed there is a need to simplify the quickstarts - to use SDKs instead of Dapr API. I also share this idea. I think we should have a simple gateway quickstarts repository, to expedite user's adoption.
> > In this PR I also provide a full list of language-specific SDKs where we have more advanced examples.
> 
> Changing the quickstarts from using APIs to SDKs must be agreed within the maintainers since that is a change in direction we have decided until now.
> 
> @AaronCrawfis @orizohar @yaron2 @msfussell Thoughts?

We should definitely not replace the HTTP examples with SDK examples.

The HTTP APIs are the means by which Dapr guarantees language agnosticism and they are consistent across languages.

SDK quickstarts can be added to complement these.

Something like

hello-world
hello-world-go-sdk
etc.

**Comment by dmitsh** [comment]
My main argument for having basic SDK quickstarts in a single repository is that having them in multiple repositories is simply confusing.
This dapr/quickstarts repo is a logical place for all types of quickstarts. We still can have more advanced examples in the language-specific SDK repos.
I agree with @yaron2.  I would also suggest making one step further and renaming hello-world into hello-world-api.

**Comment by artursouza** [comment]
> My main argument for having basic SDK quickstarts in a single repository is that having them in multiple repositories is simply confusing.
> This dapr/quickstarts repo is a logical place for all types of quickstarts. We still can have more advanced examples in the language-specific SDK repos.
> I agree with @yaron2. I would also suggest making one step further and renaming hello-world into hello-world-api.

I would not rename the hello world sample since it is already linked from multiple places.

Regarding having the quickstarts reproduced per SDK in this repo, it would require to have all of them done in a single change (IMO). Having them partially implemented will be more confusing since it would give the impression that some SDKs don't exist or are partially implemented. I think this effort is not small enough to just be closed with this PR and would need to have an issue created and prioritized within the other work items we have.

**Comment by dmitsh** [comment]
> Regarding having the quickstarts reproduced per SDK in this repo, it would require to have all of them done in a single change (IMO). Having them partially implemented will be more confusing since it would give the impression that some SDKs don't exist or are partially implemented. I think this effort is not small enough to just be closed with this PR and would need to have an issue created and prioritized within the other work items we have.

I personally don't see any issue with gradually adding hello-world-sdk quickstarts. IMO, use of SDK will significantly simplify and expedite Dapr adoption. We can start with one SDK quickstart, and add more later. 
Regarding "having them partially implemented will be more confusing since it would give the impression that some SDKs don't exist or are partially implemented", how do the developers know which language is supported today? They should somehow find the right SDK repository. If we have a centralized quickstart repo (which I'm advocating for), we can provide list and links to all SDK repos. I've already done that in this PR, but we can relocate that info where it will be more visible.

**Comment by msfussell** [comment]
I agree that SDKs simplify Dapr adoption, the issue is that SDK maintainers/approvers need to be better at creating quickstart samples in  their repo  and most have not done this enough today. The approach here is that each SDK has its own quickstarts. The discoverability issue is easily solved in docs, where you can then navigate either from here https://docs.dapr.io/ or https://docs.dapr.io/getting-started/ for the respective SDK. This cannot be done until there are better samples in the SDK.  Given there are 7 SDK (5 crafted SDKs) we should certainly not push these all into one repo. IMO this particular example is best moved into a the /examples folder in the GO SDK repo although we could also consider a quickstart repo in the Go SDK, which has a tutorial like walkthrough.  This is also agreeing the quickstart content.

**Comment by codeputer** [comment]
Personally, I totally agree.  As much as everybody wants to be inclusive across a host of languages, or to hand craft an HTTP message each time a remote request is needed, I'm NOT interested!

Give me a client side SDK that simplifies the transfer, I've simply got to much to do to learn about a 100 different choices, from a 100 different languages....

Rant over...  :)

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by dmitsh** [comment]
@msfussel, I think we all agree that SDK is the quickest way to implement Dapr application. Not having SDK example in dapr/quickstarts defies the purpose of this repo. Also, as the name suggests, it is not a quickstarts for API only, it is a quickstarts for entire Dapr.

I'm not suggesting to move all quickstarts from the language-specific SDK repos into this repository, but having some basic SDK examples, and referring to the SDK repos for more advanced cases.

**Comment by orizohar** [comment]
> @msfussel, I think we all agree that SDK is the quickest way to implement Dapr application. Not having SDK example in dapr/quickstarts defies the purpose of this repo. Also, as the name suggests, it is not a quickstarts for API only, it is a quickstarts for entire Dapr.
> 
> I'm not suggesting to move all quickstarts from the language-specific SDK repos into this repository, but having some basic SDK examples, and referring to the SDK repos for more advanced cases.
> 
> I also think that having SDK examples in different languages in this repo will serve as another proof of Dapr adoption and support within open source community.

We try to keep the quickstart repo general so it is relevant to developers using any language and at the same time small so it can be maintained and verified on each release. If we add SDKs we will need to add several quickstarts and at that point it will be grow too large in my opinion. I think this is a great addition that belongs in the go-sdk Dapr repo and I strongly disagree that not having SDK examples here "defies the purpose if this repo".
I think the maintainers can reopen this decision of having SDK specific examples in the SDK specific repos but it requires more discussion.

**Comment by dmitsh** [comment]
Let's agree to disagree :)  Closing this PR.

**Comment by AaronCrawfis** [comment]
Thanks @dmitsh for the contribution, I think having a go quickstart that mirrors the HTTP quickstart is a great idea.

I think that in a perfect implementation we have the quickstarts implementations side by side in the docs.dapr.io website, with a tabbed UI for code, showing that HTTP and SDKs are co-equal and users can choose between them.

For now though, I think we can PR this into the go SDK, and once the docs/quickstarts maintainers have time to standardize the quickstarts experience we can make it unified.

**Comment by codeputer** [comment]
Happy to agree – but I do like learning from history.  WCF was an awesome tech, but it was trying to be too much for too many people…

It feels that DAPR is also doing the same thing in terms of trying to be too much too many people.

My own personnel opinion is that .Net would be the implementation of choice, but that seems to be only one of many that is being focused on.

One of the reasons why I’m still waiting to introduce it to my clients – as IMO – DAPR has a bit of too many things to get good at any one thing.

But as you say – we can agree to disagree

**Pull request closed without merging**